src/cancellation_token.c

```diff
@@ -0,0 +1,212 @@
+
+static void cancellation_token_dispose(CANCELLATION_TOKEN* token)
+{
+    TCALL_DISPATCHER_ASSIGN(CANCELLATION_TOKEN_CANCEL_CALL)(&token->cancel_notification_call_dispatcher, NULL);
+}
+
+static void cancellation_token_registration_dispose(CANCELLATION_TOKEN_REGISTRATION* registration)
+{
+    CANCELLATION_TOKEN* token_ptr = THANDLE_GET_T(CANCELLATION_TOKEN)(registration->token);
+    TCALL_DISPATCHER_UNREGISTER_TARGET(CANCELLATION_TOKEN_CANCEL_CALL)(token_ptr->cancel_notification_call_dispatcher, registration->cancel_notification_target_handle);
+    THANDLE_ASSIGN(CANCELLATION_TOKEN)(&registration->token, NULL);
+}
+
+IMPLEMENT_MOCKABLE_FUNCTION(, THANDLE(CANCELLATION_TOKEN), cancellation_token_create, bool, canceled)
+{
+    /*Codes_SRS_CANCELLATION_TOKEN_04_001: [ cancellation_token_create shall allocate memory for a THANDLE(CANCELLATION_TOKEN). ]*/
+    THANDLE(CANCELLATION_TOKEN) token = THANDLE_MALLOC(CANCELLATION_TOKEN)(cancellation_token_dispose);
+    if (token == NULL)
+    {
+        LogError("THANDLE_MALLOC(CANCELLATION_TOKEN)(cancellation_token_dispose) failed.");
+    }
+    else
+    {
+        CANCELLATION_TOKEN* token_ptr = THANDLE_GET_T(CANCELLATION_TOKEN)(token);
+
+        /*Codes_SRS_CANCELLATION_TOKEN_04_003: [ cancellation_token_create shall create a TCALL_DISPATCHER handle by calling TCALL_DISPATCHER_CREATE(CANCELLATION_TOKEN_CANCEL_CALL). ]*/
+        TCALL_DISPATCHER(CANCELLATION_TOKEN_CANCEL_CALL) dispatcher = TCALL_DISPATCHER_CREATE(CANCELLATION_TOKEN_CANCEL_CALL)();
+
+        if (dispatcher == NULL)
+        {
+            LogError("TCALL_DISPATCHER_CREATE(CANCELLATION_TOKEN_CANCEL_CALL)() failed.");
+            TCALL_DISPATCHER_INITIALIZE(CANCELLATION_TOKEN_CANCEL_CALL)(&token_ptr->cancel_notification_call_dispatcher, NULL);
+        }
+        else
+        {
+            if (canceled == false)
+            {
+                /*Codes_SRS_CANCELLATION_TOKEN_04_004: [ cancellation_token_create shall set the initial state to be equal to canceled parameter. ]*/
+                (void)interlocked_exchange(&token_ptr->state, CANCELLATION_TOKEN_STATE_NOT_CANCELED);
+            }
+            else
+            {
+                /*Codes_SRS_CANCELLATION_TOKEN_04_004: [ cancellation_token_create shall set the initial state to be equal to canceled parameter. ]*/
+                (void)interlocked_exchange(&token_ptr->state, CANCELLATION_TOKEN_STATE_CANCELED);
+            }
+
+            TCALL_DISPATCHER_INITIALIZE_MOVE(CANCELLATION_TOKEN_CANCEL_CALL)(&token_ptr->cancel_notification_call_dispatcher, &dispatcher);
+
+            /*Codes_SRS_CANCELLATION_TOKEN_04_005: [ cancellation_token_create shall return a valid THANDLE(CANCELLATION_TOKEN) when successful. ]*/
+            goto all_ok;
+        }
+
+        /*Codes_SRS_CANCELLATION_TOKEN_04_002: [ If any underlying error occurs cancellation_token_create shall fail and return NULL. ]*/
+        THANDLE_ASSIGN(CANCELLATION_TOKEN)(&token, NULL);
```

Comment:
> THANDLE_ASSIGN(CANCELLATION_TOKEN)(&token, NULL); [](http://example.com/codeflow?start=7&length=50)

what has been done by THANDLE_MALLOC it shall be undone by THANDLE_FREE...

Reply:
In this case `token` needs to be set to `NULL` before exiting the function. If I call `THANDLE_FREE` then `token` continues to be non-`NULL`. I see that `THANDLE_ASSIGN` does a `DEC_REF` when the RHS is `NULL`. So the freeing still happens as far as I can tell.

Reply:
Ok. Added another `result` local variable and that handles it.